mgr/dashboard: fix lifecycle issues

Pegonzal opened the pull request:
Fixes: https://tracker.ceph.com/issues/68434

- Now lifecycles properly display multiple rules, before only one rule was being shown even if the lifecycle had multiple rules
- Now JSON formatted lifecycles are properly applied without the need to include 'LifecycleConfiguration' as the first element of the dictionary

[screen-capture (2).webm](https://github.com/user-attachments/assets/51f81f06-7125-4dd1-bcbb-6cd09e6a4b83)

updated video using carbon's `code-snippet` showing a lifecycle with 20 rules

[screen-capture (3).webm](https://github.com/user-attachments/assets/eb46b2c4-7be5-4ea6-9100-6e9e7a19ec7d)



--------

Regarding the code fixes:
- Main issue was due to the `Rule:` key being repeated in the json, which is not allowed, therefore only the last `Rule` was showing up, fixed it with : ` @staticmethod
    def _handle_rules(pairs):` in the backend and `replaceKey` paramter in the frontend parsing.
- Also there were some issues with the json parsing, so added `\n`


<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

afreen23 reviewed (commented):
@Pegonzal we need to consider using this component for showing json/xml https://angular.carbondesignsystem.com/?path=/docs/components-code-snippet--docs/
As I see right now with very long rules this affects visibility of other fields as it gets quite long, instead of tweaking something ourselves we can use carbon component.

Thanks

Pegonzal commented:
> @Pegonzal we need to consider using this component for showing json/xml https://angular.carbondesignsystem.com/?path=/docs/components-code-snippet--docs/ As I see right now with very long rules this affects visibility of other fields as it gets quite long, instead of tweaking something ourselves we can use carbon component.
> 
> Thanks

Thanks @afreen23, I've updated the PR with the new carbon component. Updated the description with a new video showing how it looks for a lifecycle with 20 rules

afreen23 commented:
Thanks for quick fix, one thing I noticed that scroll bar is off, not sure if we are having some custom css here.
As in carbon demo I see its perfectly aligned right.
https://carbondesignsystem.com/components/code-snippet/usage/

[Screencast from 2024-10-08 14-42-32.webm](https://github.com/user-attachments/assets/51ccf03d-e408-4bbc-9bb5-002f223b2855)

Pegonzal commented:
> Thanks for quick fix, one thing I noticed that scroll bar is off, not sure if we are having some custom css here. As in carbon demo I see its perfectly aligned right. https://carbondesignsystem.com/components/code-snippet/usage/
> 
>  Screencast.from.2024-10-08.14-42-32.webm

If I make my screen smaller I can see it properly aligned:
![image](https://github.com/user-attachments/assets/d82796b5-c7ff-494e-aac5-153dc595c875)

So looks like an issue with the responsiveness of the component?
![image](https://github.com/user-attachments/assets/7a121d38-21d0-4e9a-973e-355ff5dc53ef)

nizamial09 reviewed (approved):
lgtm, works locally. btw, you probably need to take care of the unit tests. i saw that it was failing for issues related to the changes in this PR.

Pegonzal commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-details/rgw-bucket-details.component.html:
```diff
@@ -158,8 +158,10 @@
-                  <pre *ngIf="lifecycleFormat === 'xml'">{{ (selection.lifecycle | xml) || '-'}}</pre>
+                  <cds-code-snippet display="multi">
+                    <pre *ngIf="lifecycleFormat === 'json'">{{selection.lifecycle | json}}</pre>
+                    <pre *ngIf="lifecycleFormat === 'xml'">{{ (selection.lifecycle | xml:{'Rules':'Rule'}) || '-'}}</pre>
```
Updated it

afreen23 commented:
> > Thanks for quick fix, one thing I noticed that scroll bar is off, not sure if we are having some custom css here. As in carbon demo I see its perfectly aligned right. https://carbondesignsystem.com/components/code-snippet/usage/
> > Screencast.from.2024-10-08.14-42-32.webm
> 
> If I make my screen smaller I can see it properly aligned:

Adding `width: fit-content` makes it take available space.
Lets add that and fix this meanwhile we should raise this with carbon team as well.
As we see there is some space (purple area)  between the code block and copy .

Thanks

cloudbehl commented:
jenkins test docs
